### DISCUSSION TITLE
Book recommendation about Kestrel architecture

### DISCUSSION THREAD

**Discussion by lukasgmt** [dotnet/AspNetCore.Docs#28068] [Q&A]
Could you recommend any deep dive book that describes the internal Kestrel architecture. Most of the books focus on the ASP.NET, MVC, Razor etc. while I am mostly focused on all the available middlewares and their configurations. 

What I am attempting to write is a reverse proxy that:
* Is a mitigator of OAuth authentication
* Hides OAuth tokens behind cookies
* Refresh them inside my reverse proxy depending on how recently the tokens were resolved (in a timer)
* Store tokens in Redis in order to share them across replicas of the reversed proxy

I am trying to use embedded Cookie and Authentication middlewares but so far I am failing in properly configuring them.

Would you recommend to implement it on my own or try to reuse some of OAuth + Cookies middeware?

**Comment by Rick-Anderson** [comment]
@Tratcher  can you recommend anything?

**Reply by Tratcher** [reply]
You're not looking for the internals of Kestrel, but the AspNetCore fundamentals of middleware, pipelines, requests, auth, etc. Our docs are the best reference I'm aware of for those individual pieces, but I don't think we cover more involved scenarios like this.

As for writing a reverse proxy, have you seen [YARP](https://github.com/microsoft/reverse-proxy)? While it doesn't do this for you by default, it does handle the basic proxy operations and gives you a framework to work in. I'm aware of various 3rd parties that have built something similar using YARP: https://github.com/DuendeSoftware/BFF/tree/main/src/Duende.Bff.Yarp
